src/core/angularfire2.ts

```diff
@@ -65,3 +67,36 @@ export const runInZone = (zone: NgZone) => <T>(obs$: Observable<T>): Observable<
     );
   });
 }
+
+//SEMVER: once we move to TypeScript 3.6, we can use these to build lazy interfaces
+/*
+  type FunctionPropertyNames<T> = { [K in keyof T]: T[K] extends Function ? K : never }[keyof T];
+  type PromiseReturningFunctionPropertyNames<T> = { [K in FunctionPropertyNames<T>]: ReturnType<T[K]> extends Promise<any> ? K : never }[FunctionPropertyNames<T>];
+  type NonPromiseReturningFunctionPropertyNames<T> = { [K in FunctionPropertyNames<T>]: ReturnType<T[K]> extends Promise<any> ? never : K }[FunctionPropertyNames<T>];
+  type NonFunctionPropertyNames<T> = { [K in keyof T]: T[K] extends Function ? never : K }[keyof T];
+
+  export type PromiseProxy<T> = { [K in NonFunctionPropertyNames<T>]: Promise<T[K]> } &
+    { [K in NonPromiseReturningFunctionPropertyNames<T>]: (...args: Parameters<T[K]>) => Promise<ReturnType<T[K]>> } &
+    { [K in PromiseReturningFunctionPropertyNames<T>   ]: (...args: Parameters<T[K]>) => ReturnType<T[K]> };
+*/
+
+export const _lazySDKProxy = (klass: any, observable: Observable<any>, zone: NgZone) => new Proxy(klass, {
+  get: (_, name) => zone.runOutsideAngular(() =>
+    klass[name] || new Proxy(() => 
+      observable.toPromise().then(mod => {
+        const ret = mod[name];
+        // TODO move to proper type guards
+        if (typeof ret == 'function') {
+          return ret.bind(mod);
+        } else if (ret && ret.then) {
+          return ret.then((res:any) => zone.run(() => res));
+        } else {
+          return zone.run(() => ret);
+        }
+      }), {
+        get: (self, name) => self()[name],
```

Comment:
Please bear with my ignorance, shouldn't it be `get: (self, name) => self().then(it => it[name])`, since `self()` returns a promise?

Reply:
Probably unneeded but the get here is a passthrough onto the attributes of the promise, say you wanted to get `.constructor` or `.then` without calling the function. My suspicion here is that this will be useful for type-checking or async/await.

Triggering `apply` when a function is called is why I went higher order on the first arg to the proxy.

However, I'm planning on recursing the get (and apply) here so: `self()[name] || _lazySDKProxy(klass, from(self()), zone)` which would allow chaining: `await angularFirestore.collection('asdf').ref` rather than `await angularFirestore.collection('asdf').then(c => c.ref)`